pkg/compact/schedule_delete.go

```diff
@@ -0,0 +1,143 @@
+// Licensed under the Apache License 2.0.
+
+package compact
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+// DebugCompactorMetas is folder to store compactor metadata
+// about when the block is scheduled to be deleted.
+const DebugCompactorMetas = "debug/compactor-metas"
+
+// CompactorMeta stores block id and when block was marked for deletion.
+type CompactorMeta struct {
+	ID           ulid.ULID `json:"id"`
+	DeletionTime int64     `json:"deletion_time"`
+}
+
+// ScheduleBlockDelete marks the block to be deleted.
+type ScheduleBlockDelete struct {
+	dir         string
+	logger      log.Logger
+	deleteDelay time.Duration
+	bkt         objstore.Bucket
+}
+
+// NewScheduleBlockDelete creates a new ScheduleBlockDelete.
+func NewScheduleBlockDelete(logger log.Logger, dir string, bkt objstore.Bucket, deleteDelay time.Duration) *ScheduleBlockDelete {
+	return &ScheduleBlockDelete{
+		dir:         dir,
+		logger:      logger,
+		deleteDelay: deleteDelay,
+		bkt:         bkt,
+	}
+}
+
+// ScheduleDelete deletes blocks from bucket
+// deleteDelay duration after block is marked for deletion.
+func (s *ScheduleBlockDelete) ScheduleDelete(ctx context.Context) error {
+	compactorMetaPath := filepath.Join(s.dir, "download")
+	defer func() {
+		if err := os.RemoveAll(compactorMetaPath); err != nil {
+			level.Error(s.logger).Log("msg", "failed to remove compactor meta dir", compactorMetaPath, "err", err)
+		}
+	}()
+
+	if err := os.RemoveAll(compactorMetaPath); err != nil {
+		return errors.Wrap(err, "clean compaction compactor meta dir")
+	}
+
+	if err := objstore.DownloadDir(ctx, s.logger, s.bkt, DebugCompactorMetas, compactorMetaPath); err != nil {
+		return errors.Wrap(err, "downlad compactor-metas dir")
+	}
+
+	return filepath.Walk(compactorMetaPath, func(path string, fileInfo os.FileInfo, err error) error {
+		_, file := filepath.Split(path)
+		if _, err := ulid.Parse(strings.TrimSuffix(file, filepath.Ext(file))); err != nil {
+			compactorMetaBytes, err := ioutil.ReadFile(path)
+			if err != nil {
+				return errors.Wrap(err, "read compactor meta")
+			}
+
+			compactorMeta := CompactorMeta{}
+
+			if err := json.Unmarshal([]byte(compactorMetaBytes), &compactorMeta); err != nil {
+				return errors.Wrap(err, "unmarshal compactor meta")
+			}
+
+			if time.Now().Unix()-compactorMeta.DeletionTime > s.deleteDelay.Milliseconds() {
+				if err := block.Delete(ctx, s.logger, s.bkt, compactorMeta.ID); err != nil {
+					return errors.Wrap(err, "delete block")
+				}
+
+				if err := os.RemoveAll(filepath.Join(s.dir, compactorMeta.ID.String())); err != nil {
+					return errors.Wrap(err, "delete compactor-meta.json")
+				}
```

Comment:
This is deleting local copy of the block, right? (Comment may be helpful here)

What about deletion-marker file in `DebugCompactorMetas`, shouldn't that be deleted as well, so that next run of `ScheduleDelete` doesn't find it anymore?

Reply:
the block is not downloaded locally, we just download `compactor-meta.json` file in our local machine.
After that, we check if the block can be deleted(if it was marked for deletion some time ago) then we delete the block.

I am not deleting file from `DebugCompactorMetas` in bucket because that can be used to gain information about when the block was marked for deletion.

Reply:
I didn't mean download, but delete. I was just making sure I understand the code here.